[FLINK-7552] [FLINK-7553] Enhance SinkInterface / Use in FlinkKafkaProducer

aljoscha opened the pull request:
## What is the purpose of the change

Enhance `SinkFunction` with a way of retrieving the element timestamp. This allows us to get rid of the hybrid nature of `FlinkKafkaProducer010`.

This is keeping the legacy static "convenience" methods à la `FlinkKafkaProducer010.writeToKafkaWithTimestamps` for backwards compatibility.

## Brief change log

  - Enhance Sink interface
  - Use new interface in Kafka Producer

## Verifying this change

This change is already covered by existing tests, such as *(please describe tests)*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes, call stack of KafkaProducer with writing timestamps is changed slightly, also, `StreamSink` operator now has a context object.
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

aljoscha commented:
R: @rmetzger @pnowojski Could you please review this?

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/SinkFunction.java:
```diff
@@ -35,6 +35,40 @@
+		 */
+		long timestamp();
+	}
+
```
nit/side question: do we have new line between closing braces `}` at the end of the class in the code style?

aljoscha replied:
I think our "code style" is "whatever is not rejected by our checkstyle", which is quite loose. 😉 But you're right, I'm removing the line.

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamSink.java:
```diff
@@ -31,14 +31,24 @@
 	public void processElement(StreamRecord<IN> element) throws Exception {
-		userFunction.invoke(element.getValue());
+		sinkContext.element = element;
+		userFunction.invoke(sinkContext, element.getValue());
```
wouldn't it be better/simpler to just pass `StreamRecord` to the `userFunction`?

`userFunction.invoke(element)`?

and instead of adding `SinkContext` as a first argument of the `invoke` method in the sink interface, just change the element type from `IN` to `StreamRecord<IN>`?

aljoscha replied:
`StreamRecord` is an internal concept that should not be exposed in any user facing API. That's the reason for the extract context. Plus, the context allows us to extend the information that we pass to the `SinkFunction` in the future, similar to the context in `ProcessFunction` and `ProcessWindowFunction`.

pnowojski replied:
I get it, however I still do not like this objects repacking thing done here. Instead of passing `StreamRecord` directly, we could introduce `UserRecord` interface with `getTimestamp()` method, that would be implemented by `StreamRecord`.

But that's minor complain, if you have stronger preferences for keeping it as it is, it's fine for me.

aljoscha commented:
To be honest, I didn't look at the tests because I didn't think they would fail. Sorry for that. 😓 I'll fix the tests and let you know once this is ready again.

aljoscha commented:
@rmetzger the binary compatibility checker is acting up because it doesn't like the changes to `SinkFunction` and `RichSinkFunction`. It might be that we can't actually change this. What do you think?

aljoscha commented:
The problem is that `RichSinkFunction` defines
```
public abstract void invoke(IN value) throws Exception;
```

even though it implements `SinkFunction`, which defines:
```
void invoke(IN value) throws Exception;
```

What this means is that code that is invoking these methods is different, depending on the type of reference. For example, in
```
SinkFunction<String> interfaceSink = new SinkFunction<String>() {
    @Override
    public void invoke(String value) throws Exception {

    }
};

SinkFunction<String> interfaceSinkToRichSink = new RichSinkFunction<String>() {
    @Override
    public void invoke(String value) throws Exception {

    }
};

RichSinkFunction<String> richSink = new RichSinkFunction<String>() {
    @Override
    public void invoke(String value) throws Exception {

    }
};

interfaceSink.invoke("hello");
interfaceSinkToRichSink.invoke("hello");
richSink.invoke("hello");
```
the first two calls are `invokeinterface` while the last call is using `invokevirtual`.

I don't think that people are using the sink interfaces for directly calling methods on them but this is the reason the binary compatibility checker is complaining.

rmetzger commented:
I see. It is a similar problem we've discussed with the SourceContext. Users only implement it, but don't call it.

I believe in this case, it makes sense to add an exception to Japicmp to allow this change.

aljoscha commented:
Can we add fine-grained exclusions, or would this exclude the whole file?

rmetzger commented:
No, you can define fine-grained exclusions. Check the excludes in the example: http://siom79.github.io/japicmp/MavenPlugin.html

aljoscha commented:
It seems we completely have to ignore `SinkFunction`, because japicmp will check its serialiVersionUID, which changed by adding a new method to the interface. (And yes, I know that checking the serialVersionUID on an interface does not make a lot of sense.)

pnowojski commented on flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducerBase.java:
```diff
@@ -311,7 +311,6 @@ public void invoke(IN next) throws Exception {
 		producer.send(record, callback);
 	}
 
-	@Override
```
Why did you remove `@Override`?

aljoscha replied:
That must have happened by accident, fixing...

EronWright commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/SinkFunction.java:
```diff
@@ -35,6 +35,39 @@
+	 * @param value The input record.
+	 * @throws Exception
+	 */
+	default void invoke(SinkContext context, IN value) throws Exception {
```
Consider re-ordering the two arguments to more closely match the similar signature in `ProcessFunction`.

```
default void invoke(IN value, SinkContext ctx) throws Exception
```

aljoscha replied:
Makes sense, I will change this. 👍

EronWright commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/SinkFunction.java:
```diff
@@ -35,6 +35,39 @@
+		/**
+		 * Returns the timestamp of the current input record.
+		 */
+		long timestamp();
```
Consider returning a `Long` to better match the `Context` interface in `ProcessFunction`, and document the behavior when the time characteristic is `ProcessingTime`.
```
		/**
		 * Timestamp of the element currently being processed or timestamp of a firing timer.
		 *
		 * <p>This might be {@code null}, for example if the time characteristic of your program
		 * is set to {@link org.apache.flink.streaming.api.TimeCharacteristic#ProcessingTime}.
		 */
		public abstract Long timestamp();
```

pnowojski replied:
I would be against unnecessary boxing whenever it is possible. If calling `timestamp()` doesn't make sense for processing time, then it would be better to just throw an exception.

EronWright replied:
@pnowojski let us not throw an exception here, given the commented-out section of `StreamRecord::getTimestamp` which suggests a problem with that approach.   It would be good to know why `ProcessFunction` used a `Long` rather than `long`.   I have an almost unhealthy desire for consistency.

aljoscha replied:
@StephanEwen What do you think about this? You introduced the code in `StreamRecord` that returns `Long.MIN_VALUE` when there is no timestamp instead of throwing an exception.

EronWright commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/SinkFunction.java:
```diff
@@ -35,6 +35,39 @@
+	 * @param <T> The type of elements accepted by the sink.
+	 */
+	@Public // Interface might be extended in the future with additional methods.
+	interface SinkContext<T> {
```
Looking at `ProcessFunction`, the `Context` is an abstract class and has a simpler name.  For the sake of consistency, consider the same.

aljoscha replied:
I'd like that as well but the approach won't work here because `SinkFunction` is an interface, i.e. I cannot have a `SinkFunction<T>.Context` because I can't use the generic parameter here.

aljoscha commented:
@pnowojski and @EronWright I changed the name of the `Context`, I added a test in `StreamSinkOperatorTest`, I added methods or querying current processing time/watermark to the context. I changed the `timestamp()` method to return a primitive `long` and I added a method `hasTimestamp()`. Also, `timestamp()` now throws an exception if no timestamp is available.

What do you think?

StephanEwen commented:
+1 from my side to merge it like this (using boxed long, dropping the `hasTimestamp()` method.

My reasoning is that this is consistent with `ProcessFunction` (like @EronWright said), it is also not in a hot loop and lazily created, so I do not expect a big performance hit.

@aljoscha and me talked a lot about the whole design of when records have timestamps and whether in the future we should just assume that records always have timestamps. That biased my towards dropping the `hasTimestamp()` method.

EronWright commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/SinkFunction.java:
```diff
@@ -31,10 +31,50 @@
+	}
+
+	/**
+	 * Context that {@link SinkFunction SinkFunctions } can use for getting additional data about
```
Is the link in this comment well-formed?

EronWright commented:
+1

aljoscha commented:
Thanks for the reviews, everyone! 😃
